### ISSUE TITLE
Pipelayer 0.4.1 crashes Factorio 0.18.21

### ISSUE THREAD

**Issue by diamond-lizard** [mspielberg/factorio-pipelayer#4]
While changing the input/output mode of an underground pipe connector, Factorio crashed with this error:

46325.048 Error MainLoop.cpp:1202: Exception at tick 2921485: The mod Pipelayer (0.4.1) caused a non-recoverable error.
Please report this error to the mod author.

```
Error while running event pipelayer::on_tick (ID 0)
LuaFluidBox API call when LuaFluidBox was invalid.
stack traceback:
     __pipelayer__/Connector.lua:61: in function 'ready_as_input'
     __pipelayer__/ConnectorSet.lua:98: in function 'next_input'
     __pipelayer__/Network.lua:375: in function 'try_to_transfer'
     __pipelayer__/Network.lua:394: in function 'update'
     __pipelayer__/Network.lua:357: in function 'task'
     __pipelayer__/lualib/Scheduler.lua:11: in function 'on_tick'
     __pipelayer__/Network.lua:477: in function 'update_all'
     __pipelayer__/Editor.lua:466: in function 'on_tick'
     __pipelayer__/control.lua:50: in function <__pipelayer__/control.lua:50>
     traceback:
     [C]: in function '__index'
     __pipelayer__/Connector.lua:61: in function 'ready_as_input'
     __pipelayer__/ConnectorSet.lua:98: in function 'next_input'
     __pipelayer__/Network.lua:375: in function 'try_to_transfer'
     __pipelayer__/Network.lua:394: in function 'update'
     __pipelayer__/Network.lua:357: in function 'task'
     __pipelayer__/lualib/Scheduler.lua:11: in function 'on_tick'
     __pipelayer__/Network.lua:477: in function 'update_all'
     __pipelayer__/Editor.lua:466: in function 'on_tick'
     __pipelayer__/control.lua:50: in function <__pipelayer__/control.lua:50>
```

I am using pipelayer 0.4.1 on Factorio 0.18.21, on Gentoo Linux.

**Comment by mspielberg** [comment]
Are you using Picker Dollies or perhaps a mod that does automatic deconstruction?

**Comment by diamond-lizard** [comment]
Yes, I am using Picker Dollies.

**Comment by diamond-lizard** [comment]
I am also using the Auto Deconstruct mod.

**Comment by diamond-lizard** [comment]
I just tried disabling both Picker Dollies and Auto Deconstruct, loaded up my saved game and still got the same error when I tried to change the mode of an underground pipe.

**Comment by mspielberg** [comment]
If you moved an underground pipe connector at any time using Picker Dollies, then that connector is permanently damaged. If you post a save I’ll try to fix it.

**Comment by diamond-lizard** [comment]
I've actually never used Picker Dollies.. I just had it installed because it seemed like a useful mod, but never got around to using it.

[Here](https://github.com/mspielberg/factorio-pipelayer/files/4547062/011-0216.-.fix.underground.pipes.by.oil.processing.zip) is a saved game where the issue can be reproduced by repeatedly changing the mode of some of the underground pipe connectors directly to the player's North (ie. above the player).  There seems to be some randomness involved, so it's impossible to predict exactly which connections will need to have their modes changed to trigger the issue.  Also, I've sometimes triggered the issue by deleting pipes in the underground editor.

I myself was able to work around the issue by deleting underground pipe connections and recreating them, and setting their modes before connecting any other pipes to them.  It was very painful and took a long time to do, but eventually I got it, and now I always set the mode of the underground pipe connections I create before connecting any pipes to them, and that seems to avoid triggering the bug.

Fair warning about this save: I use a lot of mods.

**Comment by mspielberg** [comment]
It seems that one of your connectors was destroyed, but not by player mining, damage, or a robot, so I presume there was some other mod that did it. I've fixed this case in v0.4.2.
